File: module-10-firewalls-and-nat/nat_firewall_sim.js
/**
 * Module 10 — JS/Node port of nat_firewall_sim.py. Same PAT table, same
 * stateful connection tracking, same 4 scenarios.
 */

const EXTERNAL_IP = "203.0.113.7";

class NatFirewall {
  constructor(externalIp = EXTERNAL_IP) {
    this.externalIp = externalIp;
    this.nextExternalPort = 40000;
    this.natTable = new Map();   // `${ip}:${port}` -> externalPort
    this.reverseNat = new Map(); // `${extIp}:${extPort}` -> [ip, port]
    this.connState = new Map(); // `${ip}:${port}:${destIp}:${destPort}` -> state
  }

  outbound(internalIp, internalPort, destIp, destPort) {
    const key = `${internalIp}:${internalPort}`;
    const connKey = `${internalIp}:${internalPort}:${destIp}:${destPort}`;

    let extPort;
    if (!this.natTable.has(key)) {
      extPort = this.nextExternalPort++;
      this.natTable.set(key, extPort);
      this.reverseNat.set(`${this.externalIp}:${extPort}`, [internalIp, internalPort]);
    } else {
      extPort = this.natTable.get(key);
    }

    const isNew = !this.connState.has(connKey);

    console.log(`  OUTBOUND ${internalIp}:${internalPort} -> ${destIp}:${destPort}`);
    console.log(`    NAT:      translated source to ${this.externalIp}:${extPort}`);
    console.log(`    firewall: connection state = ${isNew ? "NEW (creating tracked entry)" : "already tracked"}`);

    if (isNew) {
      this.connState.set(connKey, "ESTABLISHED");
    }
    return [this.externalIp, extPort];
  }

  inbound(srcIp, srcPort, destIp, destPort) {
    console.log(`  INBOUND  ${srcIp}:${srcPort} -> ${destIp}:${destPort}`);

    const natKey = `${destIp}:${destPort}`;
    if (!this.reverseNat.has(natKey)) {
      console.log(`    NAT:      no entry for ${destIp}:${destPort} -- nowhere to translate this to`);
      console.log(`    firewall: DROPPED -- unsolicited inbound traffic with no matching internal host`);
      return null;
    }

    const [internalIp, internalPort] = this.reverseNat.get(natKey);
    const connKey = `${internalIp}:${internalPort}:${srcIp}:${srcPort}`;

    if (this.connState.get(connKey) !== "ESTABLISHED") {
      console.log(`    NAT:      would translate to ${internalIp}:${internalPort}, but...`);
      console.log(`    firewall: DROPPED -- no ESTABLISHED connection state matches this source`);
      return null;
    }

    console.log(`    NAT:      translated destination to ${internalIp}:${internalPort}`);
    console.log(`    firewall: ALLOWED -- matches ESTABLISHED connection state`);
    return [internalIp, internalPort];
  }
}

function main() {
  const nf = new NatFirewall();

  console.log("=".repeat(70));
  console.log("SCENARIO 1: two internal hosts browsing the web -- PAT in action");
  console.log("=".repeat(70));
  nf.outbound("192.168.1.10", 51000, "93.184.216.34", 443);
  console.log();
  nf.outbound("192.168.1.20", 51000, "93.184.216.34", 443);
  console.log();
  console.log("  Notice: both internal hosts used the SAME internal port (51000,");
  console.log("  a coincidence that happens constantly in practice) but got DIFFERENT");
  console.log(`  external ports, both behind the one external IP ${nf.externalIp}.`);
  console.log("  This is PAT: many internal (ip, port) pairs, one external IP,");
  console.log("  disambiguated entirely by external port.");

  console.log();
  console.log("=".repeat(70));
  console.log("SCENARIO 2: the web server's reply comes back -- allowed");
  console.log("=".repeat(70));
  nf.inbound("93.184.216.34", 443, nf.externalIp, 40000);

  console.log();
  console.log("=".repeat(70));
  console.log("SCENARIO 3: a random unsolicited inbound connection -- dropped");
  console.log("=".repeat(70));
  nf.inbound("198.51.100.99", 12345, nf.externalIp, 8080);
  console.log("  Nobody inside your network ever contacted 198.51.100.99, so there's");
  console.log("  no NAT entry and no connection state for this -- this is why port-");
  console.log("  scanning a home router from the internet finds almost everything");
  console.log("  closed: there's no 'server' listening, just a translation table that");
  console.log("  only has entries for conversations YOUR network started.");

  console.log();
  console.log("=".repeat(70));
  console.log("SCENARIO 4: an inbound packet claiming an existing port, wrong source");
  console.log("=".repeat(70));
  nf.inbound("198.51.100.99", 9999, nf.externalIp, 40000);
  console.log("  Same external port as Scenario 1's real entry, but from a DIFFERENT");
  console.log("  source than the one that entry's connection was established with --");
  console.log("  still dropped. The firewall checks the FULL connection tuple, not");
  console.log("  just 'is this port open in the NAT table.'");
}

main();

File: module-10-firewalls-and-nat/nat_firewall_sim.py
"""
Module 10 -- NAT (specifically PAT, Port Address Translation) and a
stateful firewall's connection tracking, simulated together, because
that's how they actually run on your real home router: one combined
pipeline, not two separate features.

This directly answers "why can my whole house share ONE public IP" (PAT --
every internal host's connections get the SAME external IP, differentiated
by port) and "why can't something from the internet just reach my laptop
uninvited" (stateful tracking -- inbound traffic is only allowed if it
matches a connection state your own network initiated first).
"""

EXTERNAL_IP = "203.0.113.7"  # your home router's public IP (RFC 5737 example range)


class NatFirewall:
    def __init__(self, external_ip=EXTERNAL_IP):
        self.external_ip = external_ip
        self.next_external_port = 40000
        self.nat_table = {}     # (internal_ip, internal_port) -> external_port
        self.reverse_nat = {}   # (external_ip, external_port) -> (internal_ip, internal_port)
        self.conn_state = {}    # (internal_ip, internal_port, dest_ip, dest_port) -> "NEW" | "ESTABLISHED"

    def outbound(self, internal_ip, internal_port, dest_ip, dest_port):
        """A packet leaving your LAN, headed to the internet."""
        key = (internal_ip, internal_port)
        conn_key = (internal_ip, internal_port, dest_ip, dest_port)

        if key not in self.nat_table:
            ext_port = self.next_external_port
            self.next_external_port += 1
            self.nat_table[key] = ext_port
            self.reverse_nat[(self.external_ip, ext_port)] = key
        else:
            ext_port = self.nat_table[key]

        is_new = conn_key not in self.conn_state
        self.conn_state[conn_key] = "NEW" if is_new else self.conn_state[conn_key]

        print(f"  OUTBOUND {internal_ip}:{internal_port} -> {dest_ip}:{dest_port}")
        print(f"    NAT:      translated source to {self.external_ip}:{ext_port}")
        print(f"    firewall: connection state = {'NEW (creating tracked entry)' if is_new else 'already tracked'}")

        if is_new:
            self.conn_state[conn_key] = "ESTABLISHED"  # the reply we expect will match this
        return self.external_ip, ext_port

    def inbound(self, src_ip, src_port, dest_ip, dest_port):
        """A packet arriving from the internet, addressed to your public IP."""
        print(f"  INBOUND  {src_ip}:{src_port} -> {dest_ip}:{dest_port}")

        nat_key = (dest_ip, dest_port)
        if nat_key not in self.reverse_nat:
            print(f"    NAT:      no entry for {dest_ip}:{dest_port} -- nowhere to translate this to")
            print(f"    firewall: DROPPED -- unsolicited inbound traffic with no matching internal host")
            return None

        internal_ip, internal_port = self.reverse_nat[nat_key]
        conn_key = (internal_ip, internal_port, src_ip, src_port)

        if self.conn_state.get(conn_key) != "ESTABLISHED":
            print(f"    NAT:      would translate to {internal_ip}:{internal_port}, but...")
            print(f"    firewall: DROPPED -- no ESTABLISHED connection state matches this source")
            return None

        print(f"    NAT:      translated destination to {internal_ip}:{internal_port}")
        print(f"    firewall: ALLOWED -- matches ESTABLISHED connection state")
        return internal_ip, internal_port


def main():
    nf = NatFirewall()

    print("=" * 70)
    print("SCENARIO 1: two internal hosts browsing the web -- PAT in action")
    print("=" * 70)
    nf.outbound("192.168.1.10", 51000, "93.184.216.34", 443)
    print()
    nf.outbound("192.168.1.20", 51000, "93.184.216.34", 443)
    print()
    print("  Notice: both internal hosts used the SAME internal port (51000,")
    print("  a coincidence that happens constantly in practice) but got DIFFERENT")
    print(f"  external ports, both behind the one external IP {nf.external_ip}.")
    print("  This is PAT: many internal (ip, port) pairs, one external IP,")
    print("  disambiguated entirely by external port.")

    print()
    print("=" * 70)
    print("SCENARIO 2: the web server's reply comes back -- allowed")
    print("=" * 70)
    nf.inbound("93.184.216.34", 443, nf.external_ip, 40000)

    print()
    print("=" * 70)
    print("SCENARIO 3: a random unsolicited inbound connection -- dropped")
    print("=" * 70)
    nf.inbound("198.51.100.99", 12345, nf.external_ip, 8080)
    print("  Nobody inside your network ever contacted 198.51.100.99, so there's")
    print("  no NAT entry and no connection state for this -- this is why port-")
    print("  scanning a home router from the internet finds almost everything")
    print("  closed: there's no 'server' listening, just a translation table that")
    print("  only has entries for conversations YOUR network started.")

    print()
    print("=" * 70)
    print("SCENARIO 4: an inbound packet claiming an existing port, wrong source")
    print("=" * 70)
    nf.inbound("198.51.100.99", 9999, nf.external_ip, 40000)
    print("  Same external port as Scenario 1's real entry, but from a DIFFERENT")
    print("  source than the one that entry's connection was established with --")
    print("  still dropped. The firewall checks the FULL connection tuple, not")
    print("  just 'is this port open in the NAT table.'")


if __name__ == "__main__":
    main()
